version.cmake.in

```diff
@@ -0,0 +1,18 @@
+#performed in a small separate "cmake script" so can easily run on each build
+
+execute_process(
+   COMMAND @GIT_EXECUTABLE@ describe --tags --dirty
+   OUTPUT_VARIABLE VERSION_STRING
+   OUTPUT_STRIP_TRAILING_WHITESPACE
+   RESULT_VARIABLE res
+   WORKING_DIRECTORY @CMAKE_SOURCE_DIR@
+)
+if(NOT ${res} STREQUAL "0")
+  message(FATAL_ERROR "git describe failed")
+endif()
+#unsure if this next is possible but just a failsafe
+if("${VERSION_STRING}" STREQUAL "")
+  set(VERSION_STRING "unknown")
```

Comment:
Is this version of "unknown" intentionally different from the other one?

Reply:
hehe yeah very astute, to differentiate the two on failure.
Arguably there are two better approaches:
1. On newer git, use --broken which won't fail under certain submodule failure cases (like buildkite suffered from)
1. Instead of Unknown, use "Unknown (xxxx)" where xxxx is just the git hash that should have a better chance of succeeding

I'd certainly be interested in entertaining these in the future as improvements to this approach if merged.